feat: Add domain scoping to WebRetriever

vblagoje opened the pull request:
### Why:
By introducing the `allowed_domains` init parameter to `WebRetriever`, we can now limit the scope of the`WebRetriever` to a specific list of domains instead of the entire Internet. This enhancement is particularly beneficial for "talk to a website" and "talk to docs" use case scenarios, allowing for more targeted and relevant information retrieval.

### What:
Introduced a new example script, `talk_to_website.py`, that demonstrates the use of WebRetriever with various configurations and models on the use case of talking to `https://haystack.deepset.ai/` documentation. Enhanced the WebRetriever to support domain restrictions, allowing for more targeted web searches. Updated search engine providers in providers.py to support the allowed_domains parameter, enabling domain-specific searches.

### How can it be used:
This is an internal change that refines the behaviour of the WebRetriever. While it doesn't directly break the API, developers should be aware of the new domain restriction capabilities and the shift in caching behaviour.

### How did you test it:
Manually only

### Notes For Reviewer:
Please review the new example script and the modifications in the WebRetriever and search engine providers. Ensure that the domain restriction functionality works seamlessly across different providers and that the new caching logic aligns with the intended goals.

vblagoje commented:
@MichaelBitard @dfokina does `allowed_domains` or do you think there is a better name for this parameter? The goal is to limit the scope of search to a specific list of domains instead of the entire Internet. How do we call this parameter?

vblagoje commented:
Here are the candidates:

1. `target_domains`
2. `search_domains`
3. `domain_filters`
4. `domain_list`
5. `restricted_domains`
6. `allowed_domains`


I'd recommend `allowed_domains` or `search_domains` but LMK please

dfokina commented:
I would leave `allowed_domains`.
`search_domains` is great, but might confuse users in close proximity to `search_engine_providers`

vblagoje commented:
@MichelBartels Are you ok with `allowed_domains`? Can you please have a look at this one?

MichaelBitard commented:
Hi @vblagoje I think you have mistaken me for @MichelBartels , our two github handles are quite close :)

TuanaCelik commented:
> Here are the candidates:
> 
> 1. `target_domains`
> 2. `search_domains`
> 3. `domain_filters`
> 4. `domain_list`
> 5. `restricted_domains`
> 6. `allowed_domains`
> 
> I'd recommend `allowed_domains` or `search_domains` but LMK please
Personal opinion, either `allowed_domains` or `domain_filters` seem like the most appropriate. In a practical sense, as I'd be using this like a retriever, the use case of this is quite similar to metadata filtering where I'd use `filters` in a run function.. 

So my next question is, could this parameter be used in a run function if I have the node in a pipeline?

vblagoje commented:
> So my next question is, could this parameter be used in a run function if I have the node in a pipeline?

Unfortunately not, we don't have that parameter in the retrieve method of the BaseRetriever class. We can add it in the next version.

MichelBartels reviewed (changes requested):
Looks good to me except for some comments about documentation.

vblagoje commented:
@MichelBartels I've addressed your concerns in [a996cf5](https://github.com/deepset-ai/haystack/pull/5587/commits/a996cf595fe4e0ddc4a223cb9330edab57cf48ae) and [bd5c27c](https://github.com/deepset-ai/haystack/pull/5587/commits/bd5c27cc9fb9bc8b11ae3a7026ff1fd54f825bab) 
You were spot on.

MichelBartels reviewed (approved):
Looks good to me if the tests pass

silvanocerza commented:
Merged base in to fix issue with coveralls.
https://coveralls.io/builds/62270537
